proposals/20220823-huggingface_model_pusher.md

```diff
@@ -12,25 +12,48 @@
 **Project name:** HuggingFace Model Pusher
 
 ## Project Description
-HuggingFace Model Pusher(`HFModelPusher`) pushes blessed model to the [HuggingFace Model Hub](https://huggingface.co/models).
+HuggingFace Model Pusher(`HFModelPusher`) pushes blessed model to the [HuggingFace Model Hub](https://huggingface.co/models). Also, it optionally pushes application to [HuggingFace Space Hub](https://huggingface.co/spaces).
 
 ## Project Category
 Component
 
 ## Project Use-Case(s)
-The HuggingFace Model Hub lets us have [Git-LFS](https://git-lfs.github.com) enabled repositories in public and private modes. Supported models hosted on the HuggingFace Model Hub can be directly loaded/used with APIs provided by [transformers](https://huggingface.co/docs/transformers/index) package. However, it is not limited. We can host arbitrary types of models too. 
+The HuggingFace Model and Space Hubs let us have [Git-LFS](https://git-lfs.github.com) enabled repositories in public and private modes, and they are easy to manage versions sepecially for those familiar with Git. 
 
-HuggingFace Model Hub is easy to manage model versions, especially for those familiar with Git.
+Supported models hosted on the HuggingFace Model Hub can be directly loaded/used with APIs provided by [transformers](https://huggingface.co/docs/transformers/index) package. However, it is not limited. We can host arbitrary types of models too. 
+
+HuggingFace Space Hub comes with free resources to host prototype applications that use machine learning models. Currently supported application frameworks are Gradio and Streamlit. It is often a good idea to host current version of the model to the Huggingface Space, so it could be interated with real world before the production deployment. 
+
+By keeping these information in mind, `HFModelPusher` let you push a trained or blessed model to the HuggingFace Model Hub within a new branch within TFX pipeline. Then, if specified, it pushes an application to the HuggingFace Space Hub by injecting the current model information into the prepared template sources.
```

Comment:
The `hf_api` has utilities for creating a PR too. Wouldn't be nice to create a PR with the newly created branch? We can, of course, parameterize that too. 

@deep-diver

Reply:
Do you have experiences?

I couldn't find any other ways but [`upload_folder()'](https://github.com/huggingface/huggingface_hub/blob/7aadc3243d9478a48e17f568d8d4950cafd11ca2/src/huggingface_hub/hf_api.py#L2204) to create PR after commit/push changes. 

Even if `upload_folder()` let us commit changes to a particular branch, it doesn't have a capability to create a new branch, so we need to create a branch beforehand. However, we have to create a new branch and push any changes to the branch to let the remote repository detect the newly created branch. So, maybe some kind of dummy files should be pushed beforehand which I want to avoid.

The ['create_pull_request()`](https://github.com/huggingface/huggingface_hub/blob/7aadc3243d9478a48e17f568d8d4950cafd11ca2/src/huggingface_hub/hf_api.py#L2731) seems like creating a PR, but it doesn't have any parameter to specify which branch the PR for.

Reply:
@gante any pointers?

Reply:
The `revision` argument in [`upload_folder()`](https://github.com/huggingface/huggingface_hub/blob/7aadc3243d9478a48e17f568d8d4950cafd11ca2/src/huggingface_hub/hf_api.py#L2204) allows us to specify a branch name. 

I'm not sure if it creates the revision when it doesn't exist, but there are workarounds for that ([e.g.](https://huggingface.co/docs/huggingface_hub/v0.9.1/en/package_reference/repository#huggingface_hub.Repository.git_checkout))

Reply:
I see. Your suggestion is to create a branch with `git_checkout()` then call `upload_folder()`, right? but `git_checkout()` creates a branch only in the locally cloned branches. It means I have to push the branch, then upload the same files again with `upload_folder()`. I hoped there is APIs like `create_pull_request()` with revision parameter, or `upload_folder()` that creates a new branch.